Replace all direct cdata access with THPVariable_Unpack

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56150 Make it illegal to directly construct _TensorBase
* #56147 Give Tensor a trivial (for now) metaclass _TensorMeta
* #56016 Use THPVariable_Unpack in python_nccl
* **#55799 Replace all direct cdata access with THPVariable_Unpack**
* #55798 Make THPVariable_Unpack work on THPVariable too
* #55797 Modify a few APIs to take/return const Tensor& instead of Tensor&

I'm going to change the implementation of cdata soon so I need to
abstract over cdata access with a function.  Additionally, many
users are casting manually casting to THPVariable to access
the member so I can remove these unsafe casts in the client code
(the implementation, of course, is still doing an unsafe cast.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712130](https://our.internmc.facebook.com/intern/diff/D27712130)

ezyang committed:
Replace all direct cdata access with THPVariable_Unpack

I'm going to change the implementation of cdata soon so I need to
abstract over cdata access with a function.  Additionally, many
users are casting manually casting to THPVariable to access
the member so I can remove these unsafe casts in the client code
(the implementation, of course, is still doing an unsafe cast.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on torch/csrc/jit/python/python_arg_flatten.cpp:
```diff
@@ -39,12 +39,6 @@ py::object cast_handle_sequence(std::vector<py::handle> objs) {
 void flatten_rec(PyObject* obj, ParsedArgs& args) {
-  auto as_variable = [](at::Tensor& tensor) // Wrap tensor as Variable
-  {
-    PyObject* wappred_obj = THPVariable_Wrap(tensor);
```
Is it ok to remove the `THPVariable_Wrap` call here? Does it has any side effect that was important?

ezyang replied:
So, this old code wrapped the tensor in a PyObject Variable, and then *immediately* unwrapped it back into a Tensor. The only side effect of this, so far as I can tell, is that it ensures that the pyobj is set. But this is pointless as we always ensure there is a valid pyobj when doing `THPVariable_wrap`. 

The code was added in #54864 without any serious code review (cc @SplitInfinity) so my guess is it was just cargo culted from somewhere else.

albanD replied:
Sounds good!

albanD reviewed (approved):
LGTM
Nice cleanup!

ezyang commented on tools/codegen/api/python.py:
```diff
@@ -883,8 +883,8 @@ def dispatch_lambda_arg(cpp_arg: Binding) -> DispatchLambdaArgument:
-            # For method's 'self', we can use 'Tensor &' and simply ignore mutability!
-            type_str = 'Tensor &'
+            # For method's 'self', we can use 'const Tensor &' and simply ignore mutability!
+            type_str = 'const Tensor &'
```
This goes straight to the generated code so no one is matching on this. (Also, you shouldn't be matching on this lol!)

albanD replied:
I know we shouldn't but I seem to remember that somewhere. Maybe in some debug code though.
But sounds good!

ezyang committed:
Update on "Replace all direct cdata access with THPVariable_Unpack"


I'm going to change the implementation of cdata soon so I need to
abstract over cdata access with a function.  Additionally, many
users are casting manually casting to THPVariable to access
the member so I can remove these unsafe casts in the client code
(the implementation, of course, is still doing an unsafe cast.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712130](https://our.internmc.facebook.com/intern/diff/D27712130)

[ghstack-poisoned]

ezyang committed:
Update on "Replace all direct cdata access with THPVariable_Unpack"


I'm going to change the implementation of cdata soon so I need to
abstract over cdata access with a function.  Additionally, many
users are casting manually casting to THPVariable to access
the member so I can remove these unsafe casts in the client code
(the implementation, of course, is still doing an unsafe cast.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712130](https://our.internmc.facebook.com/intern/diff/D27712130)

[ghstack-poisoned]

ezyang committed:
Update on "Replace all direct cdata access with THPVariable_Unpack"


I'm going to change the implementation of cdata soon so I need to
abstract over cdata access with a function.  Additionally, many
users are casting manually casting to THPVariable to access
the member so I can remove these unsafe casts in the client code
(the implementation, of course, is still doing an unsafe cast.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D27712130](https://our.internmc.facebook.com/intern/diff/D27712130)

[ghstack-poisoned]
